### PR TITLE
initial work for Aot support

### PR DESCRIPTION
**Pull request by aitboudad**

### CODE REVIEW

**Comment by kolkov** [comment]
I try to build this...
image

**Comment by aitboudad** [comment]
@kolkov make sure you are using ui-router-core `v1.0`

**Comment by kolkov** [comment]
Thanks @aitboudad, but where can I find v1.0?
I use this instructions:
https://github.com/ui-router/ng2/blob/master/CONTRIBUTING.md

**Comment by aitboudad** [comment]
```
cd ui-router-core
git checkout 1.0.1 -b 1.0
```

**Comment by kolkov** [comment]
Thanks! Done!
image

**Comment by kolkov** [comment]
I try to build my test app with this builded version:
```
// Imports for loading & configuring the in-memory web api
import 'rxjs/add/operator/toPromise';
import 'rxjs/add/operator/map';

import {Ng2StateDeclaration, UIRouterModule, UIView, Category, trace} from "ui-router-ng2";
import {NgModule, SystemJsNgModuleLoader, NgModuleFactoryLoader} from "@angular/core";
import {BrowserModule} from '@angular/platform-browser';

//import {MyRootUIRouterConfig} from "./router.config";
import {AppComponent} from "./app.component";
//import {PublicModule} from "./public/public.module";
//import {PrivateModule} from "./private/private.module";
import {MyUIRouterConfig} from "./core/router.config";
import {CoreModule} from "./core/core.module";

export let MAIN_STATES: Ng2StateDeclaration[] = [
  // The top-level app state.
  // This state fills the root <ui-view></ui-view> (defined in index.html) with the AppComponent
  { name: 'app', component: AppComponent }
];

// Enables tracing (check the console) of:
// - TRANSITION transition start, redirect, success, error, ignored
// - VIEWCONFIG ui-view component creation/destruction and viewconfig de/activation
//trace.enable(Category.TRANSITION, Category.VIEWCONFIG);

@NgModule({
  imports: [
    BrowserModule,
    UIRouterModule.forRoot({
      //states: MAIN_STATES,
      otherwise: { state: 'public.home', params: {} },
      useHash: false,
      configClass: MyUIRouterConfig
    }),
    CoreModule,
    //PublicModule,
    //PrivateModule
  ],
  declarations: [
    AppComponent
  ],
  providers: [
    // Provide a NgModule lazy loading strategy
    { provide: NgModuleFactoryLoader, useClass: SystemJsNgModuleLoader },
    { provide: MyUIRouterConfig, useClass: MyUIRouterConfig }
  ],
  bootstrap: [ UIView ]
})
export class AppModule {}
```
image

Is it ui-router error or not?

**Comment by aitboudad** [comment]
not sure if it's related, currently I'm using the typescript version not the build one and it's work as expected.

**Comment by kolkov** [comment]
No, I reinstall `ui-router-ng2-quickstart` and error disappeared.
All work fine! Thanks!
Now I want to try ng build wit AoT...

**Comment by kolkov** [comment]
I can not check the compatibility ((
My angular-cli not working today. Yesterday all is ok, but today nothing helped (
I tried all combination of versions, deinstalling all local and global node_modules etc., but nothing good happened (((
image

**Comment by kolkov** [comment]
I have found what causes this problem in my case:
If I understand correctly, they filter out the character `!` from path, but `!` is legal character in Windows path.
https://github.com/angular/angular-cli/issues/3529#issuecomment-266737336

**Comment by kolkov** [comment]
I try to build simple app from angular-cli starter.
```
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HttpModule } from '@angular/http';

import { AppComponent } from './app.component';

import { UIRouterModule } from 'ui-router-ng2';
import {MyRootUIRouterConfig} from "./router.config";

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    UIRouterModule.forRoot({
      otherwise: { state: 'app', params: {} },
      useHash: false,
      configClass: MyRootUIRouterConfig
    }),
  ],
  providers: [
    { provide: MyRootUIRouterConfig, useClass: MyRootUIRouterConfig }
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

```
import {UIRouter} from "ui-router-ng2";
import {Injectable} from "@angular/core";
declare var SystemJS;

/**
 * Create your own configuration class (if necessary) for any root/feature/lazy module.
 *
 * Pass it to the UIRouterModule.forRoot/forChild factory methods as `configClass`.
 *
 * The class will be added to the Injector and instantiate when the module loads.
 */
@Injectable()
export class MyRootUIRouterConfig {
  /** You may inject dependencies into the constructor */
  constructor(uiRouter: UIRouter) {
    // Show the ui-router visualizer
    let vis = window['ui-router-visualizer'];
    vis.visualizer(uiRouter);
  }
}
```

image

Only this code I add to starter.
```
UIRouterModule.forRoot({
      otherwise: { state: 'app', params: {} },
      useHash: false,
      configClass: MyRootUIRouterConfig
    }),
  ],
  providers: [
    { provide: MyRootUIRouterConfig, useClass: MyRootUIRouterConfig }
  ],
```

**Comment by tonimoeckel** [comment]
@kolkov did you try to use a export function call? Like this: https://github.com/angular/angular/issues/11262

**Comment by kolkov** [comment]
Is it correct? `C:\Users\Andy\AppData\Roaming\npm\node_modules\ui-router-core ->`
```
D:\!Projects\TypeScript\ui-router\ui-router-ng2>npm link ui-router-core
D:\!Projects\TypeScript\ui-router\ui-router-ng2\node_modules\ui-router-core -> C:\Users\Andy\AppData\Roaming\npm\node_modules\ui-router-core -> D:\!Projects\TypeScript\ui-router\ui-router-core

D:\!Projects\TypeScript\ui-router\ui-router-ng2>
```

**Comment by kolkov** [comment]
@tonimoeckel If I comment `UIRouterModule.forRoot` my app builds correctly
```
@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule,
    /*UIRouterModule.forRoot({
      otherwise: { state: 'app', params: {} },
      useHash: false,
      configClass: MyRootUIRouterConfig
    }),*/
  ],
  providers: [
    { provide: MyRootUIRouterConfig, useClass: MyRootUIRouterConfig }
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
image

**Comment by aitboudad** [comment]
@kolkov you should build using `ngc` or try importing typescript files from `node_modules/` 
something like
```ts
import {UIRouterModule} from ' ../node_modules/ui-router-ng2/src/ng2';
```

**Comment by kolkov** [comment]
@aitboudad I use `ng` from angular-cli. It it correct?
If I try `import { UIRouterModule } from '../node_modules/ui-router-ng2/src/ng2';
I have the same error.
image

`

**Comment by kolkov** [comment]
The same:
image

**Comment by aitboudad** [comment]
see https://github.com/aitboudad/ui-router-aot-demo

**Comment by kolkov** [comment]
Is this ok? I see some warnings...
image

**Comment by aitboudad** [comment]
yep, I'll try to fix those warning in the next few days but it should work!

**Comment by kolkov** [comment]
@aitboudad Thanks! Very good news for us! It's work!

**Comment by kolkov** [comment]
How can we rewrite resolve in Starter?
```
{
      name: 'app.bar.details', url: '/?barId', component: BarDetailsComponent,
       resolve: [
        // Inject the barList (from the parent) and find the correct
        {
          token: 'barDetail',
          deps: ['barList', Transition],
          resolveFn: (barList, trans) =>
            barList.find(item => item.id == trans.params().barId)
        }
      ]
    },
```

**Comment by kolkov** [comment]
I rewrote my great app using beta.4 and AoT.
Everything works perfectly! 
Thanks for all very much!

**Comment by qvantor** [comment]
How soon we could use ui-router without aot problems? Cause now it still a problem

**Comment by christopherthielen** [comment]
Wow, great work everyone.

I'm trying to get angular-ui-router release candidate out the door.   After that I'll give this PR the attention it deserves!

**Comment by hghammoud** [comment]
I can't get it to work. I tried all the proposed solutions in this thread. Nada...

Can you pls rehost the working demo?

**Comment by mackelito** [comment]
Also looking for the solution to this... :/

**Comment by kolkov** [comment]
@mackelito this worked for me
image

**Comment by lexigren** [comment]
@kolkov can you create some kind of demo, of how you made ui-router work with AOT? Thanks.

**Comment by mackelito** [comment]
@kolkov so you added the index.d.ts file in "/node_modules/ui-router-ng2/"? or is that in your project dir?

**Comment by tonimoeckel** [comment]
@lexigren I just created a demo with a working version of ui-router-ng2 with AoT enabled. It is based on the latest "dev" branch of @aitboudad fork (see commit: [b649735](https://github.com/aitboudad/ng2/commit/b649735209166c2bafaf4e03f62b2445aea5339b)

File: src/ng2/lazyLoadNgModule.ts [hunk 1]
```diff
@@ -41,7 +41,7 @@ export type NgModuleToLoad = string | ModuleTypeCallback;
  * - Returns the new states array
  */
 export function loadNgModule(moduleToLoad: NgModuleToLoad): (transition: Transition) => Promise<LazyLoadResult> {
-  return function(transition: Transition) {
+  return (transition: Transition) => {
```
Comment (christopherthielen) [inline src/ng2/lazyLoadNgModule.ts]: can you explain why this was necessary? I don't see any `this`

**Comment by amcdnl** [comment]
This is epic!

File: src/ng2/lazyLoadNgModule.ts (hunk 1 above)
Comment (aitboudad) [inline src/ng2/lazyLoadNgModule.ts]: To use `function` it should be exported, it will not work with AoT when it is passed to an NgModule

**Comment by longbowww** [comment]
what is the winning config for all of us angular-cli users now? unfortunately the config of @tonimoeckel wont let cli start, just giving the old error msg `Calling function 'UIRouterModule', function calls are not supported.`

(not using the aot flag)

**Comment by marshall007** [comment]
@christopherthielen can we tag a `-beta.4` release with this patch soon? As others have mentioned, this is preventing usage with `ng-cli`.

**Comment by marshall007** [comment]
@christopherthielen @aitboudad I tried compiling under `ng-cli` using https://github.com/ui-router/ng2/commit/45c73414684256f6d0e00e5ebbe4332ac31aca6a and it was initially complaining that all my `resolveFn` needed to be exported statically. Once I did that, I'm still getting:

```
ERROR in Error encountered resolving symbol values statically. Calling function 'OpaqueToken', function calls are not supported.
Consider replacing the function or lambda with a reference to an exported function,
  resolving symbol UIROUTER_ROOT_MODULE in ./node_modules/ui-router-ng2/lib/ng2/uiRouterNgModule.d.ts,
  resolving symbol makeRootProviders in ./node_modules/ui-router-ng2/lib/ng2/uiRouterNgModule.d.ts,
  resolving symbol UIRouterModule.forRoot in ./node_modules/ui-router-ng2/lib/ng2/uiRouterNgModule.d.ts,
  resolving symbol AppModule in ./src/app/app.module.ts,
  resolving symbol AppModule in ./src/app/app.module.ts
```

- Does this mean going forward, we won't be able to use anonymous functions as our `resolveFn`?
- Should AoT be working on `master` now? Am I missing something or is this still WIP?

**Comment by marshall007** [comment]
See https://github.com/angular/angular-cli/issues/3854#issuecomment-272534091

Apparently the CLI has issues resolving dependencies when nested `node_modules` directories are present in the project. In this case, I was using `npm link ui-router-ng2` to test the most recent changes. Everything worked as expected once I removed `./node_modules/ui-router-ng2/node_modules`.

**Pull request merged**